### ISSUE TITLE
COOPERATIVE BOOM demos are desync'ing with other ports

### ISSUE THREAD

**Issue by Ch0wW** [fabiangreffrath/woof#463]
Coop demos recorded with the _Boom complevel_ are systematically desyncing for one reason or another.

I have noticed both Solo and Coop **vanilla-based demos recorded with Woof 9.0.0** are properly playbacked through DSDA-DOOM, same for the **MBF** complevel. However, only Boom 2.02 demos cannot be properly playbacked, they will systematically desync at one point.

I have tried to dig a little more about this, and from what I've seen this issue is far worse than expected.

Cooperative Boom demos used in the original BOOM 2.02 sourceport doesn't sync at all in Woof 9.0.0, and same happens in DSDA-Doom. Singleplayer Boom demos seem to playback perfectly fine however.

On the other hand, cooperative DOS demos, using the official executables or MBF Woof seem to work perfectly fine.

This table may explain my thoughts a little more:

| DEMO SOURCEPORT | SOLO/COOP? | COMPLEVEL | PLAYBACKED PORT  | RESULT |
| ------------- |:-------------:| -----:| ---- | ---- |
| DOOM2 1.9 | SOLO | VANILLA | WOOF | OK |
| DOOM2 1.9 | COOP | VANILLA | WOOF | OK |
| WOOF 9.0.0  | SOLO | VANILLA | DOOM2 1.9 | OK |
| WOOF 9.0.0  | COOP | VANILLA | DOOM2 1.9 | OK |
| WOOF 9.0.0  | SOLO | VANILLA | DSDA-DOOM | OK |
| WOOF 9.0.0  | COOP | VANILLA | DSDA-DOOM | OK |
| BOOM 2.02 | SOLO | BOOM | WOOF | OK |
| BOOM 2.02 | COOP | BOOM | WOOF | DOESN'T SYNC |
| BOOM 2.02 | SOLO | BOOM | DSDA-DOOM | OK |
| BOOM 2.02 | COOP | BOOM | DSDA-DOOM | DOESN'T SYNC |
| WOOF 9.0.0  | SOLO | BOOM | BOOM 2.02 | OK |
| WOOF 9.0.0  | COOP | BOOM | BOOM 2.02 | DOESN'T SYNC |
| WOOF 9.0.0  | SOLO | BOOM | DSDA-DOOM | OK |
| WOOF 9.0.0  | COOP | BOOM | DSDA-DOOM | DOESN'T SYNC |
| MBF 2.03  | SOLO | MBF | WOOF | OK |
| MBF 2.03  | COOP | MBF | WOOF | OK |
| WOOF 9.0.0  | SOLO | MBF | MBF 2.03 | OK |
| WOOF 9.0.0  | COOP | MBF | MBF 2.03 | OK |
| WOOF 9.0.0  | SOLO | MBF | DSDA-DOOM | OK |
| WOOF 9.0.0  | COOP | MBF | DSDA-DOOM | OK |

TL;DR: Cooperative BOOM Demos are broken, but we seem to go into a very deep rabbit hole if we try to investigate more and fix this issue. I heavily suspect a change starting with MBF (and "normalized" by PRBoom) may be the cause of it.

**Comment by fabiangreffrath** [comment]
> I have tried to dig a little more about this, and from what I've seen this issue is far worse than expected.

Well, if this happens only during coop games and only for one complevel and if the issue has been inherited by MBF and then fixed by PrBoom+, then I think we might have a good chance to track it down.

**Comment by rfomin** [comment]
> Well, if this happens only during coop games and only for one complevel and if the issue has been inherited by MBF and then fixed by PrBoom+, then I think we might have a good chance to track it down.

As far as I understand, PrBoom+ does not solve this problem, so the same dilemma arises as with complevel mbf. There are very few Boom co-op demos in the DSDA archive (none?), so I think we could try to fix that.

**Comment by fabiangreffrath** [comment]
Alright, I may have misread a part of the table. However, Pr+/DSDA is our demo sync benchmark: If DSDA desyncs a demo, it is "okay" for us to desync as well. But this line remains concerning (Woof <-> DSDA sync should be maintained):

```
WOOF 9.0.0 | COOP | BOOM | DSDA-DOOM | DOESN'T SYNC

```

**Comment by fabiangreffrath** [comment]
@Ch0wW Could you make the demos available here?

**Comment by JCDorne** [comment]
There is no boom / mbf / mbf21 complevel coop demos on DSDA yet. (4shockblast have one but it's a 20sec long on a boom wad, with only 2 monsters, it doesn't desync)
Fixing this issue will allow people start making coop demos for boom :)

I have a coop ruinbros demo (recorded with boom 2.02 on DOS) that desync near the start (at 0:45).
link to the wad : [https://www.doomworld.com/idgames/levels/doom2/p-r/ruinbros](url)
[RB06P2.zip](https://github.com/fabiangreffrath/woof/files/8141113/RB06P2.zip)
The indigo player will make a "pause" (I was controlling green player), basically the desync is right after that pause.

In boom2.02, pain elementals are close. Everything is fine.
in prboom2.02, pain elementals are far and I get damaged by the HK. desync.
in prboom2.5.0, pain elementals are far and I get damaged by the HK. desync.
in woof 9.0.0, pain elementals are far and I get damaged by the HK. desync.
in dsda-doom, pain elementals are far and I get damaged by the HK. desync
in reboom2.06, pain elementals are far and I get damaged by the HK. desync
in reboom2.06dos, pain elementals are close. Everything is fine.

It look like all ports desync the same way.
Btw, it seems that only boom2.02 demos are affected by that kind of desync in coop.
Mbf coop demos seems to playback correctly on other ports. Idk about mbf21.

If you need more demos, I can record some with @Ch0wW if needed.

**Comment by fabiangreffrath** [comment]
> In boom2.02, pain elementals are close. Everything is fine. 
> in woof 9.0.0, pain elementals are far and I get damaged by the HK. desync. 
> in dsda-doom, pain elementals are far and I get damaged by the HK. desync

See, that's a case we probably won't be able to fix, or else we will desync relative to dsda.

**Comment by Ch0wW** [comment]
> See, that's a case we probably won't be able to fix, or else we will desync relative to dsda.

If DSDA (which seems to be the port you're using for demoplaybacking accuracy) has desync issues over the original port, its initial goal is then immediately biased, and **can't be trusted for true accuracy, even for one simple test case.** It happens with all ports that have Boom support, with the exception of a fork of the real Boom 2.02 port. That shouldn't be normal at all!

The biggest issue, as @JCDorne pointed out, is that the Coop demos on Boom desync since PRBoom's creation, which is absolutely problematic. If PRBoom-based ports are **ALL wrong** for only one type of demo-playbacking, how would one trust it when it wants to display records of runs based on Boom?

Besides, if there's a case where DSDA/PRBoom devs won't also fix the issue, what are we coop speedrunners supposed to do when to record boom-demos? We shouldn't record Boom maps with the MBF format since it changes the gameplay, even very slightly.

Also, while I think of it, how would be the case that solo demos are perfectly synchronized, whereas cooperative demos lead to a desync somewhere? Fixing it **SHOULDN'T break solo demos**. If it does, then we're literally in an impasse.

One solution would be for kraflab (or all port mainteners that also give support for Boom playbacking) could give some help to correct it in all cases and maintain true accuracy.

**Comment by fabiangreffrath** [comment]
@JCDorne Do you think you could record a complevel 9 multiplayer demo with PrBoom+ that desyncs in Woof but plays back fine in DSDA?

**Comment by JCDorne** [comment]
I totally understand. There is no cl9 multiplayer demos atm. No need to keep DOS executable compatibility then :)

Basically, I want to be able to record cl9/cl11 multiplayer demos. There is no source port at the moment that :
- have a stable network code (like chocolate doom)
- is boom/mbf compatible
- work in modern OS

So basically, I tried Dos executables with Dosbox... But I figured that boom multiplayer demos are desyncing with all other source port that support mbf.
So I think the best way to "fix" that is to use a new standard for multiplayer demos. (i.e using woof/dsda-doom to record cl9 multiplayer demos)
That's why I really want that "reload button fix" to be able to start recording boom multiplayer demos in woof :) (https://github.com/fabiangreffrath/woof/issues/464)

woof is the perfect candidate for cl9/cl11 multiplayer demos :)

**Comment by fabiangreffrath** [comment]
> woof is the perfect candidate for cl9/cl11 multiplayer demos :)

And that's what it's supposed to be!

What I really need though is a series of demos that demonstrate the desyncing after using the reload button, as requested in #464.

**Comment by fabiangreffrath** [comment]
https://github.com/fabiangreffrath/woof/commit/eb82b4f64dfe9f0b65ed733acc1535cb959abe6b
